MCP Security Whitepaper V2 (#141)
cosai-oasis/ws4-secure-design-agentic-systems: whitepapers/model-context-protocol-security.md

Reviewer: ### MCP-T4: Input/Instruction Boundary Distinction Failure
→ ### MCP-T4: Data/Control Boundary Distinction Failure

Reviewer: ```suggestion
      - [MCP-T4: Data/Control Boundary Distinction Failure](#mcp-t4-datacontrol-boundary-distinction-failure)
```

Reviewer: **B3:** Prompt injection is threat 11; threat 12 is MITM.

```suggestion
4. **Resource Content Poisoning** Attackers embed hidden malicious instructions within data sources (databases, documents, API responses) that MCP servers retrieve and provide to LLMs, causing the poisoned content to execute as commands when processed, effectively achieving persistent prompt injection through trusted data channels rather than direct user input. Unlike Prompt Injection (\#11): Malicious instructions are embedded in backend data sources, not user-provided prompts. Unlike Tool Poisoning (\#2): Poisons the actual data/content retrieved by tools, not the tool definitions themselves. This attack surface may be expanded with transitive or composed MCP server calls. <a id="resource-content-poisoning"></a>
```

Reviewer: **B4:** Missing `#` makes this a relative path instead of an in-page anchor.

```suggestion
      - [MCP-T1: Improper Authentication and Identity Management](#mcp-t1-improper-authentication-and-identity-management)
```

Reviewer: **B6:** Cited UpGuard piece dates bug identification to 4 June 2025 (1 May 2025 is the server release date), does not state "1,000 enterprises", and reports no confirmed cross-user viewing or malicious exploitation. Subject is the Asana MCP server.

```suggestion
* **Asana MCP server incident** (June 2025): Tenant isolation flaw in the Asana MCP server created potential cross-organization data exposure. The cited report states there was no indication the bug was exploited or that other users viewed the exposed information. ([Read more](https://www.upguard.com/blog/asana-discloses-data-exposure-bug-in-mcp-server))
```

Reviewer: **B7:** `/specification/draft/` is a moving target. Pin to `2026-07-28`

```suggestion
2. **Tool Poisoning** Malicious modification of tool metadata, configuration, or descriptors injected into clients via the tools/list method. This can cause AI agents or MCP components to invoke, trust, or execute compromised tools, potentially leading to data leaks or system compromise. As the MCP specification notes, 'descriptions of tool behavior such as annotations should be considered untrusted, unless obtained from a trusted server' ([Key Principles](https://modelcontextprotocol.io/specification/2026-07-28/index#key-principles)), making tool poisoning a recognized risk when clients connect to unvetted servers. <a id="tool-poisoning"></a>
```

Reviewer: **B7:** `/specification/draft/` is a moving target. Pin to `2026-07-28` 

```suggestion
5. **Typosquatting/Confusion Attacks** Malicious actors create MCP servers or tools with names/descriptions similar to legitimate ones, tricking clients or AI agents into invoking harmful tools due to naming confusion or LLM hallucination. The MCP specification provides guidance on making tool origins and inputs visible to users and recommends human-in-the-loop approval for tool invocations ([User Interaction Model](https://modelcontextprotocol.io/specification/2026-07-28/server/tools#user-interaction-model)), but consent fatigue—where users reflexively approve prompts without careful review—can significantly undermine these protections. <a id="typosquatting"></a>
```

Reviewer: **B8:** This lists TEEs as an unconditional organizational must. 
3.3.2 Execution isolation places TEEs / confidential containers at Level 4 only, and only where threat model, regulation, or platform supports it. Resolve toward the profile.

```suggestion
* At Level 4, run MCP servers and clients in TEEs or confidential containers and use remote attestation prior to interactions where the threat model, regulation, or platform supports it (see §3.3.2 Isolation and Sandboxing),
```

Reviewer: **E1 (cont.):** Align TOC with §1.1.1 name. Anchor follows the renamed heading.

```suggestion
    - [6.1.2 Deployment Pattern 2: Single-Tenant Hybrid](#612-deployment-pattern-2-single-tenant-hybrid)
```

Reviewer: ```suggestion
### 6.1.2 Deployment Pattern 2: Single-Tenant Hybrid
```

Reviewer: ```suggestion
    title: "DP2: Single-Tenant Hybrid"
```

Reviewer: **E2:** Appendix / TOC call this Deployment Pattern 4 and "Multi-Tenant MCP Server", while §1.1.1 and §3.3.4 use DP3 / Multi-Tenant Cloud. 

Renumber to DP3 and converge on **Multi-Tenant Cloud**. 

```suggestion
    - [6.1.3 Deployment Pattern 3: Multi-Tenant Cloud](#613-deployment-pattern-3-multi-tenant-cloud)
```

Reviewer: ```suggestion
### 6.1.3 Deployment Pattern 3: Multi-Tenant Cloud
```

Reviewer: **E2 (cont.):** Remove editing residue (also carries a fifth name, DP4a).

```suggestion
```

Reviewer: ```suggestion
```

Reviewer: **E3:** Abstract says "almost forty threats"; §3.1 counts 7 + 8 + 19 = 34. 

```suggestion
Since its emergence a year ago, MCP has rapidly established itself as the protocol for transmitting structured context between AI agents and services. Given the growing importance and attack surface of MCP and agentic systems, it is imperative that deployment specific security threats are identified and improvements are made to address the challenges and ambiguities inherent in MCP implementations. Our primary goal is to share actionable security guidance for today's MCP implementations while identifying areas where the protocol and ecosystem may need to evolve to address emerging threats. We introduce short and medium-term security implications related to MCP through the introduction of twelve core threat categories and thirty-four threats. Our taxonomy distinguishes between traditional security threats amplified by AI and MCP, and novel attack vectors. For each threat and category, we propose mitigations, defenses, and best practices for using MCP across multiple deployment scenarios including enterprise use cases. Multiple critical CVEs have been reported and incidents such as data leakage have already occurred across MCP/agentic deployments. Several examples are mentioned with links in section 2.1.
```

Reviewer: **E3:** Abstract says "almost forty threats"; §3.1 counts 7 + 8 + 19 = 34.

```suggestion
This framework organizes thirty-four threats to MCP deployments across twelve distinct categories, spanning the full technology stack—from foundational identity and access controls through AI-specific boundary failures to supply chain and operational visibility requirements. This model enables security teams to prioritize defenses based on the specific threats and attack surfaces relevant to their deployment. The taxonomy distinguishes between traditional security concerns amplified by AI mediation and novel attack vectors unique to LLM-tool interactions, providing clear guidance for implementing defense-in-depth strategies across the MCP ecosystem.
```

Reviewer: **E4:** Doubled "are".

```suggestion
* **Tier 3 \- Conventional Threats (19 Threats)**: security threats are broadly applicable or derive from legacy, infrastructure, or transport implementation decisions
```

Reviewer: **E4:** Missing space after the period (`deployments.A`).

```suggestion
These deployment patterns each carry a distinct threat model based on the trust boundaries between client and server. Threat models differ by which components are trusted or untrusted, single- vs. multi-tenant setups, and local vs. remote deployments. A deeper analysis of the security implications of each is presented in the Appendix.
```

Reviewer: **E4:** Subject/verb disagreement.

```suggestion
* Use sandboxing to limit privilege escalation attacks: running MCP servers locally requires a sandbox to prevent privilege escalation attacks
```

Reviewer: **E5:** Anchor misspells fatigue. 

```suggestion
| [MCP-T9](#mcp-t9-trust-boundary-and-privilege-design-failures) | Trust Boundary and Privilege Design Failures | [7. Overreliance on the LLM](#overreliance) | [13. Consent/User Approval Fatigue](#consent-fatigue) |  | [Secure tool design UX Design](#secure-tool-and-ux-design) |
```

Reviewer: **E6:** Expand TPA and ATPA at first occurence.

```suggestion
**Vulnerability Examples**: Prompt injection (direct and indirect), tool poisoning attack (TPA), full schema poisoning (FSP), advanced tool poisoning attack (ATPA), resource content poisoning, hidden prompt embedding (Unicode attacks), malicious message crafting
```

Reviewer: **E7:** Document approved 8 Jan 2026 and last updated 29 May 2026; copyright still says 2025.

```suggestion
Copyright © OASIS Open 2026. All Rights Reserved. This document has been produced under the process and license terms stated in the OASIS Open Project rules: [https://www.oasis-open.org/policies-guidelines/open-projects-process](https://www.oasis-open.org/policies-guidelines/open-projects-process).
```

Reviewer: **E8:** TOC says "Anti-scope"; heading says "Anti-Scope". Match the heading.

```suggestion
    - [Anti-Scope](#anti-scope)
```

Reviewer: **L1:** NVD does not support the install-count or patch-date claims. Cite Wordfence for those, and keep NVD for the CVE record.

```suggestion
* **WordPress Plugin vulnerability**: Over 100,000 sites affected by privilege escalation via MCP in AI Engine plugin, patched June 18, 2025 ([Wordfence](https://www.wordfence.com/blog/2025/06/100000-wordpress-sites-affected-by-ai-engine-critical-privilege-escalation-vulnerability/); [NVD CVE-2025-5071](https://nvd.nist.gov/vuln/detail/CVE-2025-5071))
```

Reviewer: **L6:** Relative image path breaks outside the repo (and from `whitepapers/` does not resolve to repo-root `assets/`). Use an absolute URL.

```suggestion
![Threats](https://raw.githubusercontent.com/cosai-oasis/ws4-secure-design-agentic-systems/feat/mcp-security-v2/assets/risk-triangle.svg)
```

| [MCP-T2](#mcp-t2-missing-or-improper-access-control) | Missing or Improper Access Control  |  | [9. Insecure Human-in-the-Loop](#hil) [10. Improper Multitenancy](#improper-multitenancy) | [8. Privilege Escalation](#privilege-escalation) </br> [20. Excessive Permissions/Overexposure](#excessive-permissions) | [Secure Delegation Access Control](#secure-delegation-and-access-control)  |
| [MCP-T3](#mcp-t3-input-validationsanitization-failures) | Input Validation/Sanitization Failures |  |  | [21. Command Injection](#command-injection) </br> [22. File System Exposure/Path Traversal](#path-traversal) </br> [23. Insufficient Integrity Checks](#insufficient-integrity-checks) | [Data Sanitization Guardrails](#input-and-data-sanitization-and-filtering) </br>[Sandboxing and Isolation](#sandboxing-and-isolation) </br>Explicit tool parameters, resource URIs, or server configuration replacing deprecated Roots  |

> | [MCP-T4](#mcp-t4-inputinstruction-boundary-distinction-failure) | Data/Control Boundary Distinction Failure  | [2. Tool Poisoning](#tool-poisoning) </br> [3. Full Schema Poisoning](#fsp) </br> [4. Resource Content Poisoning](#resource-content-poisoning) | [11. Prompt Injection](#prompt-injection)  | [21. Command Injection](#command-injection) | [Input Sanitization, Guardrails Context Isolation](#input-and-data-sanitization-and-filtering) |

Reviewer: ```suggestion
| [MCP-T4](#mcp-t4-data-control-boundary-distinction-failure) | Data/Control Boundary Distinction Failure  | [2. Tool Poisoning](#tool-poisoning) </br> [3. Full Schema Poisoning](#fsp) </br> [4. Resource Content Poisoning](#resource-content-poisoning) | [11. Prompt Injection](#prompt-injection)  | [21. Command Injection](#command-injection) | [Input Sanitization, Guardrails Context Isolation](#input-and-data-sanitization-and-filtering) |
```

10. **Improper Multitenancy** An attacker may exploit weak isolation between tenants or users, such as shared memory between processes, sessions,  or secrets and credentials, to access or manipulate unauthorized data. <a id="improper-multitenancy"></a>
11. **Prompt Injection** LLMs have insufficient boundaries between input data and instructions. Attackers craft malicious inputs to manipulate LLMs or MCP components to perform unintended or harmful actions such as data exfiltration, privilege escalation, or unauthorized command execution. These malicious instructions can be sent *directly* to the LLM (e.g., through a server's direct LLM provider integration or through legacy Sampling flows) or *indirectly* by embedding instructions in prompts, resources, tool metadata, tool outputs, or MCP App UI content. This threat exists whenever untrusted input can reach the LLM's context window. <a id="prompt-injection"></a>
12. **Man-in-the-Middle (MITM)** Exploiting insecure network transport (lack of TLS, improper certificate validation, or missing mutual authentication) to intercept, modify, or reroute data between MCP components, enabling data theft or manipulation. <a id="mitm"></a>

> 13. **Consent/User Approval Fatigue** Flooding users with excessive consent or permission prompts, causing habituation and leading to blind approval of potentially dangerous or malicious actions. <a id="user-fagitue"></a>

Reviewer: **E5:** Anchor misspells fatigue. 

```suggestion
13. **Consent/User Approval Fatigue** Flooding users with excessive consent or permission prompts, causing habituation and leading to blind approval of potentially dangerous or malicious actions. <a id="consent-fatigue"></a>
```

Secure identity, authentication, and authorization across the agentic and MCP ecosystem is an extremely active area of research and development. We will provide a much deeper analysis of the problem space in a subsequent white paper.

### 3.2.2 Secure Delegation and Access Control
<a id="secure-delegation-and-access-control"></a>

> To mitigate against privilege escalation, MCP servers should operate with the minimum privileges necessary. OAuth provides a widely adopted framework for secure delegation, with extensions that support fine-grained scope control and secure token flows (see [MCP Authorization](https://modelcontextprotocol.io/specification/draft/basic/authorization)).

Reviewer: **B7:** `/specification/draft/` is a moving target. Pin to `2026-07-28` 

```suggestion
To mitigate against privilege escalation, MCP servers should operate with the minimum privileges necessary. OAuth provides a widely adopted framework for secure delegation, with extensions that support fine-grained scope control and secure token flows (see [MCP Authorization](https://modelcontextprotocol.io/specification/2026-07-28/basic/authorization)).
```

* enforce the use of MCP hosts and clients with a configuration that unprivileged users cannot change and that keeps the confirmation prompt enabled.

> * use [elicitation](https://modelcontextprotocol.io/specification/draft/client/elicitation) on the MCP server side to request the user confirmation of actions.

Reviewer: **B7:** `/specification/draft/` is a moving target. Pin to `2026-07-28` 

```suggestion
* use [elicitation](https://modelcontextprotocol.io/specification/2026-07-28/client/elicitation) on the MCP server side to request the user confirmation of actions.
```

* **Schema and errors:** Tool schemas now use JSON Schema 2020-12 by default and can express richer validation logic. Implementations must be robust against expensive schemas, schema poisoning, and unsafe external references. Resource-not-found handling should be updated from the older MCP custom `-32002` code to JSON-RPC `-32602` (`Invalid Params`).

## 3.3 Security Assurance Profiles

The preceding sections define twelve threat categories and their controls and mitigations. A flat list of controls, however, does not help practitioners determine which controls apply to their specific deployment. A sandbox experiment and a multi-tenant production system do not share the same threat surface. Applying uniform requirements causes friction: sandbox experiments stall under heavy compliance burdens, while production systems carry hidden risk due to under-investment.

> Assurance profiles map specific security controls to specific deployment contexts. This section defines four assurance levels for MCP deployments. Each level specifies concrete control requirements across eight security dimensions, maps back to the MCP-T1 through MCP-T12 threat categories defined above (with cross-references to the [OWASP MCP Top 10](https://owasp.org/www-project-mcp-top-10/)), and aligns with the deployment patterns in Appendix 6.1. The [2026 MCP roadmap](https://blog.modelcontextprotocol.io/posts/2026-mcp-roadmap/) identifies enterprise readiness as a top priority; these profiles provide the security dimension of that effort.

Reviewer: **L3:** OWASP MCP Top 10 is cited with no version or retrieval date, and the target is pre-release (v0.1 / Version 0.0.0, Phase 3 beta). 

```suggestion
Assurance profiles map specific security controls to specific deployment contexts. This section defines four assurance levels for MCP deployments. Each level specifies concrete control requirements across eight security dimensions, maps back to the MCP-T1 through MCP-T12 threat categories defined above (with cross-references to the [OWASP MCP Top 10](https://owasp.org/www-project-mcp-top-10/) version [v0.1], OWASP Incubator Phase 3 beta, retrieved 30 July 2026), and aligns with the deployment patterns in Appendix 6.1. The [2026 MCP roadmap](https://blog.modelcontextprotocol.io/posts/2026-mcp-roadmap/) identifies enterprise readiness as a top priority; these profiles provide the security dimension of that effort.
```

### 3.3.1 Level Definitions

| Level | Alias | Intended Use | Trust Assumptions |
| :---- | :---- | :---- | :---- |
| Level 1 | Sandbox | Local development, single-user experiments, proof-of-concept work | Single user, no sensitive data, failures are recoverable. No access to production credentials or live data. |
| Level 2 | Internal | Internal enterprise deployments, team-shared environments, staging | Authenticated users within an organization. Moderate blast radius. Incidents must be diagnosable after the fact. |
| Level 3 | Production | Production workloads handling sensitive or business-critical data | Active threat surface. Data loss or unauthorized access has material business impact. Full auditability required. |
| Level 4 | Regulated | Multi-tenant platforms, adversarial environments, regulatory scope | Active adversaries assumed. Regulatory audit obligations. Strong tenant isolation and workload attestation required. |

### 3.3.2 Control Requirements by Level

#### Identity and Authentication

| Control | L1 | L2 | L3 | L4 |
| :---- | :---- | :---- | :---- | :---- |

> | MCP client-server authentication | Implicit (local process via stdio) | MUST: [OAuth 2.1](https://modelcontextprotocol.io/specification/draft/basic/authorization) with PKCE for all remote server connections | MUST: OAuth 2.1 with PKCE, short-lived credentials, and [Client ID Metadata Documents](https://modelcontextprotocol.io/specification/draft/basic/authorization#client-registration-approaches) where supported | MUST: OAuth 2.1 with PKCE, short-lived credentials, and enterprise-managed authorization where applicable |

Reviewer: B7: /specification/draft/ is a moving target. Pin to 2026-07-28

```suggestion
| MCP client-server authentication | Implicit (local process via stdio) | MUST: [OAuth 2.1](https://modelcontextprotocol.io/specification/2026-07-28/basic/authorization) with PKCE for all remote server connections | MUST: OAuth 2.1 with PKCE, short-lived credentials, and [Client ID Metadata Documents](https://modelcontextprotocol.io/specification/2026-07-28/basic/authorization/client-registration#client-id-metadata-documents) where supported | MUST: OAuth 2.1 with PKCE, short-lived credentials, and enterprise-managed authorization where applicable |
```

| Credential storage | Local config acceptable. Real credentials discouraged; if production or user credentials are present, reclassify as Level 2. | MUST: OS keychain or secrets manager | MUST: secrets manager with automated rotation policy | MUST: long-lived private keys for signing and identity should be protected by hardware-backed or isolated keystores where available. Ephemeral access tokens must remain short-lived and runtime-confined. |
| Credential lifetime | Real credentials discouraged. If used, treat as Level 2. | SHOULD: bounded lifetime, rotation on schedule | MUST: short-lived tokens (minutes to hours); refresh tokens rotated and stored as confidential credentials | MUST: short-lived sender-constrained credentials, typically minutes, bounded by transaction or task risk |

**Threat coverage:** MCP-T1 (Identity/Auth Failures), MCP-T2 (Authorization Errors)
**OWASP MCP mapping:** MCP01 (Token Mismanagement), MCP07 (Insufficient Auth)

#### Authorization and Delegation

| Control | L1 | L2 | L3 | L4 |
| :---- | :---- | :---- | :---- | :---- |

> | Tool-level authorization | Not required | SHOULD: tool allowlists per agent role | MUST: ABAC/PBAC for tool access, supplemented by tool-based access control (TBAC) enforcing parameter-level constraints. TBAC evaluates the specific tool request against verifiable agent identity before execution. | MUST: TBAC with continuous real-time evaluation, anomaly detection, and automated revocation of compromised agent credentials |

Reviewer: **E6:** Expand PBAC at first use.

```suggestion
| Tool-level authorization | Not required | SHOULD: tool allowlists per agent role | MUST: attribute-based access control (ABAC) / policy-based access control (PBAC) for tool access, supplemented by tool-based access control (TBAC) enforcing parameter-level constraints. TBAC evaluates the specific tool request against verifiable agent identity before execution. | MUST: TBAC with continuous real-time evaluation, anomaly detection, and automated revocation of compromised agent credentials |
```

> | Scope management | No restriction | SHOULD: scoped permissions per server connection | MUST: least-privilege scopes; [incremental scope consent](https://modelcontextprotocol.io/specification/draft/basic/authorization#scope-challenge-handling) via `WWW-Authenticate` challenges rather than requesting maximal scopes upfront; clients preserve previously granted scopes during step-up | MUST: structured authorization intent via [Rich Authorization Requests (RFC 9396)](https://datatracker.ietf.org/doc/html/rfc9396) using the `authorization_details` parameter for all sensitive tool operations |

Reviewer: B7: /specification/draft/ is a moving target. Pin to 2026-07-28

```suggestion
| Scope management | No restriction | SHOULD: scoped permissions per server connection | MUST: least-privilege scopes; [incremental scope consent](https://modelcontextprotocol.io/specification/2026-07-28/basic/authorization#scope-challenge-handling) via `WWW-Authenticate` challenges rather than requesting maximal scopes upfront; clients preserve previously granted scopes during step-up | MUST: structured authorization intent via [Rich Authorization Requests (RFC 9396)](https://datatracker.ietf.org/doc/html/rfc9396) using the `authorization_details` parameter for all sensitive tool operations |
```

| Audience restriction | Not required | SHOULD: audience (`aud`) claim validation on received tokens | MUST: [Resource Indicators (RFC 8707)](https://datatracker.ietf.org/doc/html/rfc8707) specifying the target MCP server's canonical URI. Authorization server must reject tokens where audience does not match. | MUST: Resource Indicators enforced globally, integrated with dynamic endpoint discovery via [Protected Resource Metadata (RFC 9728)](https://datatracker.ietf.org/doc/rfc9728/) |
| Token passthrough | Not applicable | MUST: MCP servers must not accept tokens intended for themselves and forward those same tokens to upstream APIs. See [MCP security best practices](https://modelcontextprotocol.io/specification/2025-06-18/basic/security_best_practices#token-passthrough). | MUST: downstream authentication managed exclusively via [token exchange (RFC 8693)](https://datatracker.ietf.org/doc/html/rfc8693) producing distinct, scoped tokens for each upstream resource | MUST: token exchange with sender-constrained bindings for all downstream calls |

> | Human-in-the-loop | Warning on untrusted tools | MUST: explicit user confirmation for all destructive or state-mutating actions. MCP [elicitation](https://modelcontextprotocol.io/specification/draft/client/elicitation) and `InputRequiredResult` provide request-scoped mechanisms for server-requested user input. | MUST: configurable centralized approval policies with step-up authentication for high-impact actions; approval prompts traceable to the initiating request | MUST: multi-party approval workflows for irreversible or high-value operations, with continuous context validation |

Reviewer: B7: /specification/draft/ is a moving target. Pin to 2026-07-28

```suggestion
| Human-in-the-loop | Warning on untrusted tools | MUST: explicit user confirmation for all destructive or state-mutating actions. MCP [elicitation](https://modelcontextprotocol.io/specification/2026-07-28/client/elicitation) and `InputRequiredResult` provide request-scoped mechanisms for server-requested user input. | MUST: configurable centralized approval policies with step-up authentication for high-impact actions; approval prompts traceable to the initiating request | MUST: multi-party approval workflows for irreversible or high-value operations, with continuous context validation |
```

| Monitoring and alerting | Not required | SHOULD: centralized log aggregation | MUST: SIEM integration, anomaly detection on agent behavior | MUST: continuous monitoring, automated containment triggers, incident response hooks (kill-switch, tool disablement) |

**Threat coverage:** MCP-T12 (Logging Gaps)
**OWASP MCP mapping:** MCP08 (Lack of Audit and Telemetry)

#### Supply Chain and Lifecycle

| Control | L1 | L2 | L3 | L4 |
| :---- | :---- | :---- | :---- | :---- |
| Server provenance | Warning when running unverified servers | SHOULD: provenance checks for tool definitions and server packages | MUST: code signing verification before installation, SBOM tracking | MUST: code signing, SBOM, reproducible builds, binary authorization. Execution should be blocked for any server lacking a signed SBOM verified within a defined window. |

> | Server inventory | Not required | SHOULD: documented inventory of deployed servers with owner and trust status. The [MCP Registry](https://modelcontextprotocol.io/specification/2025-11-25) provides a discovery mechanism for known servers. | MUST: centralized inventory with metadata (version, owner, purpose, allowed deployment environments) | MUST: centralized inventory with automated discovery of shadow deployments. Production environments must detect and alert on unregistered MCP servers. |

Reviewer: **B5:** `https://modelcontextprotocol.io/specification/2025-11-25` is the spec overview, not the registry. 

```suggestion
| Server inventory | Not required | SHOULD: documented inventory of deployed servers with owner and trust status. The [Official MCP Registry](https://registry.modelcontextprotocol.io/) provides a discovery mechanism for known servers. | MUST: centralized inventory with metadata (version, owner, purpose, allowed deployment environments) | MUST: centralized inventory with automated discovery of shadow deployments. Production environments must detect and alert on unregistered MCP servers. |
```

| Cacheable result handling | Not required | SHOULD: honor `ttlMs` and `cacheScope`, defaulting to private/no-store when uncertain | MUST: prevent cross-user and cross-tenant caching; revalidate pinned definitions after expiry or list-change notifications | MUST: centralized cache policy with security event invalidation and tenant-isolated cache keys |
| SSRF and traversal defense | Not required | MUST: restrict target URL schemes to HTTPS. Sanitize all file paths against directory traversal. | MUST: deny DNS resolution to loopback (127.0.0.0/8), link-local, and cloud metadata endpoints (169.254.169.254) | MUST: all network egress through strict allowlists managed by inspected egress proxies |

**Threat coverage:** MCP-T3 (Input Validation Failures), MCP-T4 (Data/Control Boundary Failures)

> **OWASP MCP mapping:** MCP03 (Tool Poisoning), MCP05 (Command Injection), MCP06 (Prompt Injection via Context)

Reviewer: **L3:** Paper says "Prompt Injection via Context"; OWASP uses "Intent Flow Subversion" and "Prompt Injection via Contextual Payloads". 

```suggestion
**OWASP MCP mapping:** MCP03 (Tool Poisoning), MCP05 (Command Injection), MCP06 (Prompt Injection via Contextual Payloads)
```

| Explicit state handle integrity | Not required | MUST: transport session identifiers must not be used as authenticators. Verify authorization on every request and scope explicit handles to the authenticated principal. | MUST: explicit handles, task IDs, and continuation state are tenant-bound, expiring, revocable, and logged. Stateful MCP servers must verify authorization per request. | MUST: handle integrity with continuous re-evaluation and automated revocation on anomaly detection |
| Server and authorization discovery | Not applicable (local) | SHOULD: support `server/discover` and [OAuth Protected Resource Metadata (RFC 9728)](https://datatracker.ietf.org/doc/rfc9728/) for server and authorization server discovery | MUST: MCP servers implement `server/discover` for protocol versions/capabilities and expose Protected Resource Metadata for authorization server discovery. Clients should support incremental scope elevation via `WWW-Authenticate` challenges. | MUST: `server/discover`, Protected Resource Metadata, and [OIDC Discovery](https://openid.net/specs/openid-connect-discovery-1_0-final.html) validation with centralized policy enforcement |

> | Elicitation security | Not required | SHOULD: treat server-provided prompts and [elicitation requests](https://modelcontextprotocol.io/specification/draft/client/elicitation) as untrusted. Servers must not use form mode to request credentials. | MUST: `InputRequiredResult` and task `input_required` flows isolated from privileged tool invocation unless policy explicitly allows coupling and logs it | MUST: elicitation-originated content treated as untrusted input requiring full validation before influencing tool parameters or execution |

Reviewer: B7: /specification/draft/ is a moving target. Pin to 2026-07-28

```suggestion
| Elicitation security | Not required | SHOULD: treat server-provided prompts and [elicitation requests](https://modelcontextprotocol.io/specification/2026-07-28/client/elicitation) as untrusted. Servers must not use form mode to request credentials. | MUST: `InputRequiredResult` and task `input_required` flows isolated from privileged tool invocation unless policy explicitly allows coupling and logs it | MUST: elicitation-originated content treated as untrusted input requiring full validation before influencing tool parameters or execution |
```

| Refresh tokens | Not applicable | If issued, must be protected as confidential credentials and should be rotated | MUST: refresh tokens rotated, stored securely, and not treated as substitutes for runtime authorization at the MCP server | MUST: refresh token rotation enforced, with revocation propagated within defined SLA |

**Threat coverage:** MCP-T1 (Identity/Auth Failures), MCP-T7 (Session/Transport Failures)
**OWASP MCP mapping:** MCP01 (Token Mismanagement), MCP07 (Insufficient Auth)

### 3.3.3 Threat Coverage Summary

This matrix shows the minimum expected coverage per threat category at each level. [OWASP MCP Top 10](https://owasp.org/www-project-mcp-top-10/) categories are cross-referenced where applicable.

| CoSAI Threat Category | OWASP MCP Top 10 | L1 | L2 | L3 | L4 |
| :---- | :---- | :---- | :---- | :---- | :---- |
| MCP-T1: Identity/Auth Failures | MCP01, MCP07 | Partial | MUST | MUST | MUST |
| MCP-T2: Authorization Errors | MCP02, MCP07 | Partial | MUST | MUST | MUST |

> | MCP-T3: Input Validation Failures | MCP05 | SHOULD | MUST | MUST | MUST |

Reviewer: **B9:** Coverage matrix disagrees with 3.3.2 Tool Definition, Input, and Output Integrity: Level 2 schema validation and output handling are MUST there, but MCP-T4 Level 2 is SHOULD here. 
OWASP mapping also disagrees: this row’s sibling MCP-T3 lists MCP05 only, while that dimension maps MCP03, MCP05, and MCP06. Generate this matrix from the control tables.

```suggestion
| MCP-T3: Input Validation Failures | MCP03, MCP05, MCP06 | SHOULD | MUST | MUST | MUST |
```

> | MCP-T4: Data/Control Boundary Failures | MCP03, MCP06 | Partial | SHOULD | MUST | MUST |

Reviewer: **B9:** Coverage matrix disagrees with 3.3.2 Tool Definition, Input, and Output Integrity: Level 2 schema validation and output handling are MUST there, but MCP-T4 Level 2 is SHOULD here. 
OWASP mapping also disagrees: this row’s sibling MCP-T3 lists MCP05 only, while that dimension maps MCP03, MCP05, and MCP06. Generate this matrix from the control tables.

```suggestion
| MCP-T4: Data/Control Boundary Failures | MCP03, MCP06 | Partial | MUST | MUST | MUST |
```

| MCP-T5: Data Protection | MCP10 | Partial (mock data) | MUST | MUST | MUST |
| MCP-T6: Integrity/Verification | MCP03, MCP04 | Not required | SHOULD | MUST | MUST |
| MCP-T7: Session/Transport Failures | MCP01 | SHOULD (localhost) | MUST | MUST | MUST |
| MCP-T8: Network Binding Failures | MCP09 | MUST (localhost bind) | MUST | MUST | MUST |
| MCP-T9: Trust Boundary Failures | MCP02 | Partial | MUST | MUST | MUST |
| MCP-T10: Resource Management | -- | Not required | MUST | MUST | MUST |
| MCP-T11: Supply Chain Failures | MCP04 | Not required | SHOULD | MUST | MUST |
| MCP-T12: Logging Gaps | MCP08 | Not required | MUST | MUST | MUST |

### 3.3.4 Deployment Pattern Mapping

| Deployment Pattern | Typical Level | Notes |
| :---- | :---- | :---- |
| DP1: All-Local (stdio) | Level 1 or 2 | Level 1 for personal experiments. Level 2 if handling internal data or if multiple users share the environment. |

> | DP2: Single-Tenant Remote | Level 2 or 3 | Level 2 for internal tools. Level 3 when serving production workloads or sensitive data. |

Reviewer: **E1:** DP2 is named three ways: "Single-Tenant Hybrid" (§1.1.1), "Single-Tenant Remote" (here), and "Single-Tenant MCP Server" (appendix / TOC). 

Converge on **Single-Tenant Hybrid**. 

```suggestion
| DP2: Single-Tenant Hybrid | Level 2 or 3 | Level 2 for internal tools. Level 3 when serving production workloads or sensitive data. |
```

**Token binding is a hard requirement at Level 3.** Without sender-constrained tokens, a compromised agent context, explicit state handle, or task workflow can be replayed against downstream services. Teams operating at Level 3 without [DPoP](https://datatracker.ietf.org/doc/html/rfc9449) or mTLS binding are accepting a risk that contradicts the level's stated trust assumptions.

**Level 1 is data isolation, not "no security."** Level 1 does not require runtime process sandboxing (containers, chroot), but it prohibits access to production credentials, live data, and production endpoints. If a Level 1 deployment touches real credentials, it is miscategorized. This is a classification decision, not a configuration knob.

**Level transitions should be planned, not reactive.** A proof-of-concept that succeeds at Level 1 will be pushed toward production. Upgrading from Level 2 to Level 3 requires migrating from standard OAuth tokens to sender-constrained tokens (DPoP), implementing workload identity (e.g., SPIFFE / SPIRE SVIDs), enforcing TBAC on tool execution, and enabling continuous ABAC evaluation.

> **OWASP Minimum Bar as a Level 2 prerequisite.** The [OWASP MCP Security Minimum Bar](https://genai.owasp.org/resource/a-practical-guide-for-secure-mcp-server-development/) defines strict deployment gates across five categories (identity, isolation, tooling, validation, deployment). Fulfilling those requirements is the prerequisite for achieving Level 2. Organizations that cannot pass the OWASP baseline should not deploy MCP servers in shared environments.

Reviewer: **L2:** Link text does not match the published title. Cite the guide by its real title (and name any internal section if "Minimum Bar" is meant as a subset).

```suggestion
**OWASP secure MCP server guide as a Level 2 prerequisite.** [A Practical Guide for Secure MCP Server Development](https://genai.owasp.org/resource/a-practical-guide-for-secure-mcp-server-development/) (OWASP Gen AI Security Project, 16 February 2026) defines strict deployment gates across five categories (identity, isolation, tooling, validation, deployment). Fulfilling those requirements is the prerequisite for achieving Level 2. Organizations that cannot pass that baseline should not deploy MCP servers in shared environments.
```

**Agent gateways for Level 3 and Level 4.** Embedding complex authorization, payload inspection, and identity management logic directly into individual MCP servers creates inconsistent security posture. Deployments targeting Level 3 and above should deploy a centralized proxy or agent gateway as the primary enforcement point for token validation, TBAC policy evaluation, workload identity exchange, and audit logging. Decoupling enforcement from tool execution ensures uniform compliance across heterogeneous server implementations.

> **Profiles can be verified.** Each control in the matrix is testable. Future work should define automated checks (similar to SLSA verification tooling and tools like [mcp-scan](https://github.com/invariantlabs-ai/mcp-scan)) that validate whether a deployment meets a claimed level. Continuous verification is critical: a server deployed securely on Tuesday may be compromised via an upstream dependency update by Thursday. Level 3 deployments should integrate continuous scanning into deployment pipelines. Level 4 deployments should enforce binary authorization blocking servers without current signed SBOMs.

Reviewer: **L8:** Recommends an unpinned third-party scanner, against this paper's own pinning rules. Repo has moved to Snyk; pin a release tag and state verification expectation.

```suggestion
**Profiles can be verified.** Each control in the matrix is testable. Future work should define automated checks (similar to SLSA verification tooling and tools like [mcp-scan / agent-scan v0.5.15](https://github.com/snyk/agent-scan/releases/tag/v0.5.15); verify the release signature or commit hash before use) that validate whether a deployment meets a claimed level. Continuous verification is critical: a server deployed securely on Tuesday may be compromised via an upstream dependency update by Thursday. Level 3 deployments should integrate continuous scanning into deployment pipelines. Level 4 deployments should enforce binary authorization blocking servers without current signed SBOMs.
```

### 3.3.6 Emerging Standards

The agent identity and authorization landscape is evolving rapidly. The following standards and frameworks are informative references for implementers. As these mature, future profile revisions will incorporate them where appropriate.

* [AI Agent Authentication and Authorization](https://datatracker.ietf.org/doc/draft-klrc-aiagent-auth/) (`draft-klrc-aiagent-auth-01`, Kasselman et al., March 2026): Proposes a comprehensive model using [WIMSE](https://datatracker.ietf.org/wg/wimse/about/) architecture, SPIFFE identifiers, and OAuth extensions. Standardizes agents as workloads within the WIMSE framework. Relevant to the identity controls at Levels 3 and 4.

> * [OAuth Identity and Authorization Chaining](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-identity-chaining-08) (`draft-ietf-oauth-identity-chaining-08`): Formalizes cross-domain delegation, extending [RFC 8693](https://datatracker.ietf.org/doc/html/rfc8693) token exchange for multi-hop agent chains. Relevant to scope narrowing and delegation depth controls.

Reviewer: **L4:** Internet-Drafts expire; add author and date like the `draft-klrc-aiagent-auth-01` entry.

```suggestion
* [OAuth Identity and Authorization Chaining](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-identity-chaining-08) (`draft-ietf-oauth-identity-chaining-08`, Schwenkschuster et al., 9 February 2026): Formalizes cross-domain delegation, extending [RFC 8693](https://datatracker.ietf.org/doc/html/rfc8693) token exchange for multi-hop agent chains. Relevant to scope narrowing and delegation depth controls.
```

> * [Transaction Tokens for A2A](https://datatracker.ietf.org/doc/html/draft-liu-oauth-a2a-profile-00) (`draft-liu-oauth-a2a-profile-00`): Provides an alternative to nested `act` claims for preserving call chain context in agent-to-agent communication. May address token size concerns in deep delegation chains.

Reviewer: ```suggestion
* [Transaction Tokens for A2A](https://datatracker.ietf.org/doc/html/draft-liu-oauth-a2a-profile-00) (`draft-liu-oauth-a2a-profile-00`, Liu and Yuan, 20 October 2025): Provides an alternative to nested `act` claims for preserving call chain context in agent-to-agent communication. May address token size concerns in deep delegation chains.
```

* [OAuth SPIFFE Client Authentication](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-spiffe-client-auth-00) (`draft-ietf-oauth-spiffe-client-auth-00`): Profiles SPIFFE SVIDs as OAuth client credentials, enabling agents to authenticate without client secrets. Relevant to Level 3+ credential management.

> * [AGNTCY](https://agntcy.org/) (Linux Foundation): Provides reference implementations for Agent Identity Badges and Tool-Based Access Control (TBAC). The TBAC model is referenced in the authorization controls at Level 3+. See [#47](https://github.com/cosai-oasis/ws4-secure-design-agentic-systems/issues/47) and [#48](https://github.com/cosai-oasis/ws4-secure-design-agentic-systems/issues/48) for evaluation threads.

Reviewer: **E6:** TBAC already expanded earlier; do not expand again.

```suggestion
* [AGNTCY](https://agntcy.org/) (Linux Foundation): Provides reference implementations for Agent Identity Badges and TBAC. The TBAC model is referenced in the authorization controls at Level 3+. See [#47](https://github.com/cosai-oasis/ws4-secure-design-agentic-systems/issues/47) and [#48](https://github.com/cosai-oasis/ws4-secure-design-agentic-systems/issues/48) for evaluation threads.
```

### MCP-T12: Insufficient Logging, Monitoring, and Auditability

**Technical Description**: Absence of standardized audit logging, comprehensive traceability mechanisms, and security monitoring capabilities within MCP implementations. Without robust logging of MCP server requests, request metadata, tool invocations, extension use, authorization decisions, task lifecycle events, cache decisions, and data access patterns, organizations face significant compliance blind spots and inability to perform forensic analysis of security incidents. The deprecated MCP Logging feature should not be treated as the security telemetry plane; use structured logs and OpenTelemetry instead.

> **Architectural Impact:** Severely impairs incident detection, response capabilities, and post-incident forensics. Organizations cannot trace AI agent actions back to their source, establish accountability for security breaches, or identify patterns indicating compromise or abuse.Compliance frameworks requiring audit trails and accountability mechanisms cannot be satisfied, creating regulatory risks and limiting the ability to meet security certification requirements.

Reviewer: **E4:** Missing space after `abuse.`

```suggestion
**Architectural Impact:** Severely impairs incident detection, response capabilities, and post-incident forensics. Organizations cannot trace AI agent actions back to their source, establish accountability for security breaches, or identify patterns indicating compromise or abuse. Compliance frameworks requiring audit trails and accountability mechanisms cannot be satisfied, creating regulatory risks and limiting the ability to meet security certification requirements.
```
